### PR TITLE
[FEATURE][ML] Boosted tree improvements plus more unit testing

### PR DESCRIPTION
**Pull request by tveasey**
This improves various parameter initialisations. It also improves weight shrinkage. Finally, it switches to consider both the mean loss and its variation across the folds when choosing the best hyperparameter values. I've also done a lot more work on unit testing. Although there is still some more to do I think this represents a reasonable point to review.

### CODE REVIEW

File: lib/maths/CBoostedTree.cc [hunk 1]
```diff
@@ -498,7 +501,7 @@ class CBoostedTree::CImpl final {
     }
 
     void maximumHyperparameterOptimisationRounds(std::size_t rounds) {
-        m_MaximumHyperparameterOptimisationRounds = rounds;
+        m_MaximumOptimisationRoundsPerHyperparameter = rounds;
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: Should we also rename the function?

File: lib/maths/CBoostedTree.cc [hunk 2]
```diff
@@ -531,15 +534,19 @@ class CBoostedTree::CImpl final {
...
+        for (std::size_t round = 0; round < numberRounds; ++round) {
             LOG_TRACE(<< "Optimisation round = " << round + 1);
 
             TMeanVarAccumulator lossMoments{this->crossValidateForest(
                 frame, trainingRowMasks, testingRowMasks, recordProgress)};
 
-            this->captureBestHyperparameters(CBasicStatistics::mean(lossMoments));
+            this->captureBestHyperparameters(lossMoments);
 
+            if (std::sqrt(CBasicStatistics::variance(lossMoments)) <
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: Could you add a comment why we break here?

File: lib/maths/CBoostedTree.cc [hunk 3]
```diff
@@ -708,9 +719,15 @@ class CBoostedTree::CImpl final {
         m_Gamma = m_GammaOverride.value_or(0.0);
         if (m_EtaOverride) {
             m_Eta = *m_EtaOverride;
+        } else {
+            m_Eta = 1.0 / std::max(10.0, std::sqrt(static_cast<double>(
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: A comment that explains why we init those the way we do would be great

File: lib/maths/CBoostedTree.cc [hunk 4]
```diff
@@ -708,9 +719,15 @@ class CBoostedTree::CImpl final {
...
+                                             frame.numberColumns() - 4)));
+            m_EtaGrowthRatePerTree = 1.0 + m_Eta / 2.0;
         }
         if (m_MaximumNumberTreesOverride) {
             m_MaximumNumberTrees = *m_MaximumNumberTreesOverride;
+        } else {
+            m_MaximumNumberTrees = static_cast<std::size_t>(2.0 / m_Eta + 0.5);
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: ditto

Comment (dimitris-athanasiou) [reply lib/maths/CBoostedTree.cc]: Also, this seems like it could be zero. Would that be fine?

File: lib/maths/CBoostedTree.cc [hunk 5]
```diff
@@ -757,7 +774,7 @@ class CBoostedTree::CImpl final {
         // We allow a large number of trees by default in the main parameter
         // optimisation loop. In practice, we should use many fewer if they
         // don't significantly improve test error.
-        m_MaximumNumberTrees = 500;
+        m_MaximumNumberTrees *= 10;
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: would it make sense to bound this to avoid very large numbers?

File: lib/maths/CBoostedTree.cc [hunk 6]
```diff
@@ -1154,14 +1177,23 @@ class CBoostedTree::CImpl final {
             result.emplace_back(std::log(m_Gamma / 10.0), std::log(10.0 * m_Gamma));
         }
         if (m_EtaOverride == boost::none) {
-            result.emplace_back(0.8, 1.0);
+            double rate{m_EtaGrowthRatePerTree - 1.0};
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: Could you add a comment to explain what happens here (and all the magic numbers)?

File: lib/maths/CBoostedTree.cc [hunk 7]
```diff
@@ -1226,14 +1263,18 @@ class CBoostedTree::CImpl final {
...
+        LOG_TRACE(<< "lambda* = " << m_Lambda << ", gamma* = " << m_Gamma
+                  << ", eta* = " << m_Eta << ", eta growth rate per tree* = " << m_EtaGrowthRatePerTree
+                  << ", feature bag fraction* = " << m_FeatureBagFraction);
     }
 
     //! Get the number of hyperparameters to tune.
     std::size_t numberHyperparametersToTune() const {
         return (m_LambdaOverride ? 0 : 1) + (m_GammaOverride ? 0 : 1) +
-               (m_EtaOverride ? 0 : 1) + (m_FeatureBagFractionOverride ? 0 : 1);
+               (m_EtaOverride ? 0 : 2) + (m_FeatureBagFractionOverride ? 0 : 1);
```
Comment (dimitris-athanasiou) [inline lib/maths/CBoostedTree.cc]: For eta, is the second hyperparam  its growth rate?

File: lib/maths/unittest/CBoostedTreeTest.cc [hunk 8]
```diff
@@ -114,33 +101,385 @@ void CBoostedTreeTest::testPiecewiseConstant() {
             LOG_DEBUG(<< "function moments = " << functionMoments);
             LOG_DEBUG(<< "model prediction error moments = " << modelPredictionErrorMoments);
 
-            // Unbiased...
-            CPPUNIT_ASSERT(maths::CBasicStatistics::mean(modelPredictionErrorMoments) < 0.01);
```
Comment (dimitris-athanasiou) [inline lib/maths/unittest/CBoostedTreeTest.cc]: Are we now asserting something here?

File: lib/maths/CBoostedTree.cc (hunk 5 above)
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: It is implicitly bounded by the value of the learn rate. But I think the observation does then apply to the learn rate override, i.e. we should clearly trap that this is greater than zero. I'll apply some sanity checks on the overrides.

File: lib/maths/CBoostedTree.cc (hunk 7 above)
Comment (tveasey) [reply lib/maths/CBoostedTree.cc]: Yes

File: lib/maths/unittest/CBoostedTreeTest.cc (hunk 8 above)
Comment (tveasey) [reply lib/maths/unittest/CBoostedTreeTest.cc]: This is an artefact of reworking the tests, i.e. this `CPPUNIT_ASSERT` is compared against a helper function which extracts the statistics the tests assert on.

**Comment by tveasey** [comment]
Thanks for the review @dimitris-athanasiou. Regarding bounding the number of trees I wasn't totally sure whether to impose any sanity checking on the parameter overrides at this level, but in the end I decided it is  probably worth it. This addresses, indirectly, all routes by which this could become v.large. I also added comments where you requested I think. Can you take another look?

**Pull request merged**